### PR TITLE
hw6 task2 classes and tests

### PR DESCRIPTION
**Pull request by IuliiaLanchuk**

### CODE REVIEW

File: homework6/task2_oop2/project2_oop2/oop2.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+        Return True if the deadline has not expired, False otherwise.
+        """
+        return self.created + self.deadline > datetime.datetime.now()
+
+
+class DeadlineError(Exception):
+    """Raise when homework is inactive when its deadline has already ended."""
+
+    pass
```
Comment (Cjkjvfnby) [inline homework6/task2_oop2/project2_oop2/oop2.py]: Do you need `pass` here?

File: homework6/task2_oop2/project2_oop2/oop2.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+
+class NameSurnameCreation:
```
Comment (Cjkjvfnby) [inline homework6/task2_oop2/project2_oop2/oop2.py]: Why not name it `Person`?

File: homework6/task2_oop2/project2_oop2/oop2.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+class HomeworkResult:
+    """Contain information about homework result.
+
+    Raise HomeworkError if a given homework is not a Homework class object.
+    """
+
+    def __init__(self, homework: Homework, solution: str) -> None:
+        if not isinstance(homework, Homework):
+            raise HomeworkTypeError("You gave not a Homework object")
+        self.author = Student
```
Comment (Cjkjvfnby) [inline homework6/task2_oop2/project2_oop2/oop2.py]: Do you find it strange to assign a class to an instance variable?

File: homework6/task2_oop2/project2_oop2/oop2.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        self.created = datetime.datetime.now()
+
+
+class Student(NameSurnameCreation):
+    """Create Student object."""
+
+    @staticmethod
+    def do_homework(
+        home_work: Homework, result: str
+    ) -> Union[HomeworkResult, DeadlineError]:
```
Comment (Cjkjvfnby) [inline homework6/task2_oop2/project2_oop2/oop2.py]: That does not work in that way.  You don't return DeadlineError.  There is no way to describe exceptions that could be raised by annotations.

File: homework6/task2_oop2/tests/test_oop2.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    opp_teacher = Teacher("Daniil", "Shadrin")
+    oop_hw = opp_teacher.create_homework("Learn OOP", 1)
+
+    assert oop_hw.text == "Learn OOP"
+    assert oop_hw.deadline == datetime.timedelta(days=1)
+
+
+def test_active_homework_return_homeworkresult_class_object_positive_test():
+    good_student = Student("Lev", "Sokolov")
+    opp_teacher = Teacher("Daniil", "Shadrin")
```
Comment (Cjkjvfnby) [inline homework6/task2_oop2/tests/test_oop2.py]: Why do you need a teacher instance here?

File: homework6/task2_oop2/tests/test_oop2.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    hw_inactive = advanced_python_teacher.create_homework("Read docs", 0)
+    with pytest.raises(DeadlineError, match="You are late"):
+        lazy_student.do_homework(hw_inactive, "I have done this hw too")
+
+
+def test_homework_is_not_a_homework_class_object_in_homework_result_class():
+    with pytest.raises(HomeworkTypeError, match="You gave not a Homework object"):
+
+        assert not isinstance(
+            HomeworkResult("it is not a Homework class object", "Solution").homework,
```
Comment (Cjkjvfnby) [inline homework6/task2_oop2/tests/test_oop2.py]: Do you really need to care about return value?

File: homework6/task2_oop2/tests/test_oop2.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+def test_reset_homework_was_not_given_as_an_argument_result_in_full_reset():
+    opp_teacher = Teacher("Daniil", "Shadrin")
+    good_student = Student("Lev", "Sokolov")
+    oop_hw = opp_teacher.create_homework("Learn OOP", 1)
+    result_1 = good_student.do_homework(oop_hw, "I have done this hw")
+    opp_teacher.check_homework(result_1)
+
+    assert "I have done this hw" in Teacher.homework_done[oop_hw.text]
+    assert "Learn OOP" in Teacher.homework_done
+    Teacher.reset_results()
```
Comment (Cjkjvfnby) [inline homework6/task2_oop2/tests/test_oop2.py:119]: What state of the Teacher.homework_done you expect after that?  How can you easily test it?

File: homework6/task2_oop2/tests/test_oop2.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+def test_teacher_create_homework():
+    opp_teacher = Teacher("Daniil", "Shadrin")
+    oop_hw = opp_teacher.create_homework("Learn OOP", 1)
+
+    assert oop_hw.text == "Learn OOP"
+    assert oop_hw.deadline == datetime.timedelta(days=1)
+
+
+def test_active_homework_return_homeworkresult_class_object_positive_test():
```
Comment (Cjkjvfnby) [inline homework6/task2_oop2/tests/test_oop2.py]: You can skip technical details.  Use test to speak in terms of business. `test_active_homework_is_done_by_student`

File: homework6/task2_oop2/tests/test_oop2.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+def test_inactive_homework_raise_deadline_error():
+    lazy_student = Student("Roman", "Petrov")
+    advanced_python_teacher = Teacher("Aleksandr", "Smetanin")
+    hw_inactive = advanced_python_teacher.create_homework("Read docs", 0)
+    with pytest.raises(DeadlineError, match="You are late"):
+        lazy_student.do_homework(hw_inactive, "I have done this hw too")
+
+
+def test_homework_is_not_a_homework_class_object_in_homework_result_class():
```
Comment (Cjkjvfnby) [inline homework6/task2_oop2/tests/test_oop2.py]: `test homework result constructor raises an error on wrong argument type`

File: homework6/task2_oop2/tests/test_oop2.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    opp_teacher = Teacher("Daniil", "Shadrin")
+    hw_task = opp_teacher.create_homework("Learn english", 2)
+    result_negative = bad_student.do_homework(hw_task, "< 5")
+    opp_teacher.check_homework(result_negative)
+
+    assert "Learn english" not in opp_teacher.homework_done
+    assert "< 5" not in opp_teacher.homework_done[result_negative.homework.text]
+
+
+def test_reset_homework_was_given_as_an_argument_result_in_only_this_homework_result_deletion():
```
Comment (Cjkjvfnby) [inline homework6/task2_oop2/tests/test_oop2.py]: Let's simplify a name a little.
`reset_homework deletes only given homework`

File: homework6/task2_oop2/tests/test_oop2.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    opp_teacher = Teacher("Daniil", "Shadrin")
+    good_student = Student("Lev", "Sokolov")
+    oop_hw = opp_teacher.create_homework("Learn OOP", 1)
+    result_1 = good_student.do_homework(oop_hw, "I have done this hw")
+    opp_teacher.check_homework(result_1)
+
+    assert "I have done this hw" in Teacher.homework_done[oop_hw.text]
+    assert "Learn OOP" in Teacher.homework_done
+    Teacher.reset_results(oop_hw)
```
Comment (Cjkjvfnby) [inline homework6/task2_oop2/tests/test_oop2.py]: You show your intention by using `only` that means if we change code to delete all `all` test will fail.  Will it?

```suggestion
    Teacher.reset_results()
```

File: homework6/task2_oop2/tests/test_oop2.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    good_student = Student("Lev", "Sokolov")
+    oop_hw = opp_teacher.create_homework("Learn OOP", 1)
+    result_1 = good_student.do_homework(oop_hw, "I have done this hw")
+    opp_teacher.check_homework(result_1)
+
+    assert "I have done this hw" in Teacher.homework_done[oop_hw.text]
+    assert "Learn OOP" in Teacher.homework_done
+    Teacher.reset_results()
+    assert "Learn OOP" not in Teacher.homework_done
+    assert "I have done this hw" not in Teacher.homework_done[oop_hw.text]
```
Comment (Cjkjvfnby) [inline homework6/task2_oop2/tests/test_oop2.py]:
```python
assert "undefined" is not "a function"
```
Does it prove anything?  It is better just to test state itself.

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: homework6/task2_oop2/project2_oop2/oop2.py (hunk 1 above)
Comment (IuliiaLanchuk) [reply homework6/task2_oop2/project2_oop2/oop2.py]: no, fixed

File: homework6/task2_oop2/project2_oop2/oop2.py (hunk 3 above)
Comment (IuliiaLanchuk) [reply homework6/task2_oop2/project2_oop2/oop2.py]: Fixed

File: homework6/task2_oop2/project2_oop2/oop2.py (hunk 4 above)
Comment (IuliiaLanchuk) [reply homework6/task2_oop2/project2_oop2/oop2.py]: Fixed

File: homework6/task2_oop2/tests/test_oop2.py (hunk 5 above)
Comment (IuliiaLanchuk) [reply homework6/task2_oop2/tests/test_oop2.py]: Removed

File: homework6/task2_oop2/tests/test_oop2.py (hunk 7 above)
Comment (IuliiaLanchuk) [reply homework6/task2_oop2/tests/test_oop2.py:119]: len(Teacher.homework_done) = 0

File: homework6/task2_oop2/tests/test_oop2.py (hunk 11 above)
Comment (IuliiaLanchuk) [reply homework6/task2_oop2/tests/test_oop2.py]: Test will also pass. Okey, I understand, we need to add another student with his hw result and check its presence in dictionary after Teacher.reset_results(oop_hw)

File: homework6/task2_oop2/project2_oop2/oop2.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+
+    def __init__(self, first_name: str, last_name: str):
+        self.first_name = first_name
+        self.last_name = last_name
+
+
+class Student(Person):
+    """Create Student object."""
+
+    def do_homework(self, home_work: Homework, result: str) -> Any:
```
Comment (Cjkjvfnby) [inline homework6/task2_oop2/project2_oop2/oop2.py]: Why any?  There is only one option for return value here and it is HomeworkResult

File: homework6/task2_oop2/project2_oop2/oop2.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+
+    @staticmethod
+    def reset_results(homework: Homework = None) -> None:
+        """Depending on a given argument, this method partly or fully remove results in homework_done data structure.
+
+        If homework is not given as an argument, this method remove all information in homework_done data structure.
+        Otherwise, it removes only results of a given homework.
+        """
+        if homework:
+            Teacher.homework_done[homework.text].clear()
```
Comment (Cjkjvfnby) [inline homework6/task2_oop2/project2_oop2/oop2.py]: I'd prefer using `del Teacher.homework_done[homework.text]`  both approaches are work absolutely the same but have slightly different performance under specific cases.  `del` case is more general, and in my opinion better show your intention.

File: homework6/task2_oop2/tests/test_oop2.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+
+def test_homework_result_constructor_raises_an_error_on_wrong_argument_type():
+    with pytest.raises(HomeworkTypeError, match="You gave not a Homework object"):
+        assert HomeworkResult(
+            Student("Roman", "Petrov"), "it is not a Homework class object", "Solution"
+        ).homework
+
+
+def test_teacher_check_homework_with_length_more_than_5_result_in_add_in_homework_done_data_structure():
```
Comment (Cjkjvfnby) [inline homework6/task2_oop2/tests/test_oop2.py]: `in_add_in` looks strange.

```
test check_homework saves result to homework_done
```

**Review by Cjkjvfnby — APPROVED** [review]

**Pull request merged**